lib/s3.persist.stream.js

```diff
@@ -0,0 +1,127 @@
+var AWS = require('aws-sdk');
+var _ = require('underscore');
+
+var stream = require('stream');
+var util = require('util');
+
+var rotateHistoryCount = 3;
+var PassThrough = stream.PassThrough;
+
+function S3PersistStream(options) {
+  var awsCredentials = new AWS.Credentials({accessKeyId: options.accessKeyId, secretAccessKey: options.secretAccessKey});
+  this.s3 = new AWS.S3();
+  this.s3.config.credentials = awsCredentials;
+  this.s3Stream = require('s3-upload-stream')(this.s3);
+
+  if (!(this instanceof S3PersistStream)) {
+    return new S3PersistStream(options);
+  }
+  PassThrough.call(this, options);
+
+  if (!options.metadata_version) {
+    throw new Error('Metadata Version must be specified for S3 persist stream');
+  }
+
+  if (!options.organization) {
+    throw new Error('Organization must be specified for S3 persist stream');
+  }
+
+  if (!options.stream_id) {
+    throw new Error('Stream ID must be specified for S3 persist stream');
+  }
+
+  if (!options.bucket) {
+    throw new Error('Bucket must be specified for S3 persist stream');
+  }
+
+  this._metadata_version = options.metadata_version;
+  this._organization = options.organization;
+  this._stream_id = options.stream_id;
+  this._bucket = options.bucket;
+
+  this._uploadParams = {
+    "Bucket": this._bucket,
+    "Key": util.format('%s/%s/%s', this._organization, this._stream_id, this._metadata_version),
+    "StorageClass": "REDUCED_REDUNDANCY"
+  }
+  this._listObjectParams = {Bucket: this._bucket,
+                            Prefix: util.format('%s/%s/', this._organization, this._stream_id)};
+
+  this._uploadStream = null;
+  this._writeError = null;
+}
+
+util.inherits(S3PersistStream, PassThrough);
+
+S3PersistStream.prototype._write = function (chunk, enc, done) {
+  var self = this;
+
+  if (!this._writeError && this._uploadStream == null) {
+    try {
+      this._uploadStream = self.s3Stream.upload(this._uploadParams);
+      this._uploadStream.maxPartSize(10485760);
+
+      this._uploadStream.on('error', function(e) {
+        self._writeError = e;
+        self.emit('error', self._writeError);
+      });
+
+      this._uploadStream.on('uploaded', function(details) {
+        self.rotate();
+        self.emit('persistdone', details);
+      });
+
+    } catch (e) {
+      this._writeError = e;
+      this.emit('error', this._writeError);
+    }
+
+  }
+
+  self.push(chunk);
+
+  if (!this._writeError) {
+    return this._uploadStream.write(chunk, enc, done);
```

Comment:
My comment from before (haven't seen your response): aren't we back-pressuring on the wrong stream?